ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/deduplication/CommandDeduplicationBase.scala

```diff
@@ -192,6 +193,56 @@ private[testtool] abstract class CommandDeduplicationBase(
     }
   )
 
+  testGivenAllParticipants(
+    s"${testNamingPrefix}SimpleDeduplicationMixedClients",
+    "Deduplicate commands within the deduplication time window using the command client and the command submission client",
+    allocate(SingleParty),
+    runConcurrently = false,
+  )(implicit ec =>
+    configuredParticipants => { case Participants(Participant(ledger, party)) =>
+      val submitAndWaitRequest = ledger
+        .submitAndWaitRequest(party, Dummy(party).create.command)
+        .update(
+          _.commands.deduplicationTime := deduplicationDuration.asProtobuf
+        )
+      val submitRequest = ledger
+        .submitRequest(party, Dummy(party).create.command)
+        .update(_.commands.commandId := submitAndWaitRequest.getCommands.commandId)
+      val firstSubmitAndWait = Random.nextBoolean()
+      runGivenDeduplicationWait(configuredParticipants) {
+        def submitAndAssertAccepted(submitAndWait: Boolean) = {
+          if (submitAndWait) ledger.submitAndWait(submitAndWaitRequest)
+          else submitRequestAndAssertCompletionAccepted(ledger)(submitRequest, party)
+        }
+
+        def submitAndAssertDeduplicated(submitAndWait: Boolean) = {
+          if (submitAndWait)
+            submitAndWaitRequestAndAssertDeduplication(ledger)(submitAndWaitRequest)
+          else submitRequestAndAssertDeduplication(ledger)(submitRequest, party)
+        }
+
+        deduplicationWait =>
+          for {
+            // Submit command (first deduplication window)
+            _ <- submitAndAssertAccepted(firstSubmitAndWait)
+            _ <- submitAndAssertDeduplicated(firstSubmitAndWait)
```

Comment:
This doesn't seem to test the cases where we use one endpoint, then the other. For example, submit with the CommandService, and then check that the CommandSubmissionService deduplicates, and vice versa.

I suggest we just generate 4 random booleans, and run the test 10 times.

Reply:
(Or systematically test all 16 possibilities.)

Reply:
Ok, I will just test all the possibilities in parallel